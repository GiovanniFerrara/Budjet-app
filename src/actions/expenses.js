import uuid from 'uuid';
import db from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseObj = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseObj;
    const expense = { description, note, amount, createdAt }
    return db.ref('expenses')
      .push(expense)
      .then((ref) => {
        dispatch(addExpense({
          ...expense,
          id: ref.key
        }))
      })
      .catch(e => console.log(e))
  }
}


// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
