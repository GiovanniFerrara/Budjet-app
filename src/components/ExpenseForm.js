import React, { Component } from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css'

const now = moment();
console.log(now.format("Do - MMM - YYYY"))

class ExpenseForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
        description: props.expense?  props.expense.description : "",
        createdAt: props.expense?  moment(props.expense.createdAt) : moment(),
        amount: props.expense?  (props.expense.amount/100).toString() : 0,
        note: props.expense? props.expense.note : "",
        calendarFocus: false,
        error: "",
    }
  }
  handleSubmit = (e) => {
    e.preventDefault();
    if(!this.state.description || !this.state.amount ){
      this.setState(()=>({
        error : "Please enter description and amount"
      }))
    } else {
      this.setState(()=>({
        error : ""
      }))
      this.props.onSubmit({
        description: this.state.description,
        createdAt: this.state.createdAt.valueOf(),
        amount: parseFloat(this.state.amount,10)*100,
        note: this.state.note,
      })
    }
  }
  onDateChange = (createdAt)=>{
    if(createdAt){
      this.setState(()=>({createdAt}))
    }
  }
  onFocusChange = ({focused}) => {
    this.setState(()=> ({calendarFocus: focused}))
  }
  handleInputChange = (e)=>{
    const input = e.target.name;
    const value = e.target.value;
    this.setState(()=>({
      [input]: value,
    }))
  }
  render() {
    return (
      <div>
        {this.state.error && <h4> {this.state.error} </h4>}
        <form onSubmit={this.handleSubmit}>
          <input type="text"
            placeholder="description"
            name="description"
            autoFocus
            onChange={this.handleInputChange}
            noValidate = {true}
            value = {this.state.description}
            
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocus} // PropTypes.bool
            onFocusChange={this.onFocusChange}
            id="your_unique_id" // PropTypes.string.isRequired,
            numberOfMonths={2}
            isOutsideRange = {()=> false}
          />
          <input type="number"
            name="amount"
            placeholder="Amount"
            onChange={this.handleInputChange}
            value = {this.state.amount}
          />
          <textarea type="text"
            name="note"
            placeholder="Notes"
            onChange={this.handleInputChange}
            value = {this.state.note}
          />
          <button> Add Expense </button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
