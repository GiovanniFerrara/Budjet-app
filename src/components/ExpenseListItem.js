import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

export const ExpenseListItem = ({ description, amount, id, createdAt, dispatch }) => {
  return (
    <div>
      <h3><Link to={"edit/" + id}>{description}</Link></h3>
      <p> {amount / 100} -  {moment(createdAt).format('DD-MMMM-YYYY')} </p>
    </div>
  );
}


export default ExpenseListItem;
