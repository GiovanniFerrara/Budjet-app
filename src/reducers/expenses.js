import { editExpense } from "../actions/expenses";

// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {

    case 'ADD_EXPENSE':
      return addExpense(state, action.expense);

    case 'REMOVE_EXPENSE':
      return removeExpense(state, action.id)

    case 'EDIT_EXPENSE':
      return editExpenseById(state, action.id, action.updates)

    case 'FETCH_EXPENSES':
      return setExpenses(action.expenses)

    default:
      return state;
  }
};

const setExpenses = (expenses) => {
  return expenses
}

const addExpense = (state, expenseObj) => {
  return [...state, expenseObj]
}

const removeExpense = (state, expenseId) => {
  return state.filter(({ id }) => id !== expenseId)
}

const editExpenseById = (state, id, expenseUpdatesObj) => {
  return state.map((expense) => {
    if (expense.id === id) {
      return {
        ...expense,
        ...expenseUpdatesObj
      };
    } else {
      return expense;
    };
  });
}