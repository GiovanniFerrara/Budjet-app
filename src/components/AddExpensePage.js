import React, { Component } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends Component {

  onSubmit = (expense) => {
    this.props.startAddExpense(expense)
    this.props.history.push("/dashboard")
  }
  render() {
    return (
      <div>
        This is from my add expense component
    <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => {
    dispatch(startAddExpense(expense))
  }
}
)

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
