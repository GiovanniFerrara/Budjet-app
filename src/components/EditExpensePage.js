import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses'

const EditExpensePage = ({dispatch,history,expense,match}) => (
  <div>
    This is from my edit expense component
    <ExpenseForm 
      onSubmit ={ (newExpense)=> {
        console.log(newExpense)
        dispatch(editExpense(expense.id,newExpense));
        history.push("/")
      }}
      expense = {expense}
    />
    
    <button onClick={()=>{
      dispatch(removeExpense({id: expense.id}))
      history.push("/")
    }
      }> Remove </button>
  </div>
);
const mapStateToProps = (state, props)=>({
  expense: state.expenses.find((expense)=>{
    return expense.id = props.match.params.id
  })
})
export default connect(mapStateToProps)(EditExpensePage)
