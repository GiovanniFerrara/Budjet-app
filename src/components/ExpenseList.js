import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => (
  <div>
    <h1>Expense List</h1>
    {expenses.length === 0 ?
      <p>No Expenses added still</p> :
      expenses.map((item) => (
        <ExpenseListItem {...item} key={item.id} />
      ))}
  </div>
);

const mapStateToProps = (state) => {
  return { expenses: getVisibleExpenses(state.expenses, state.filters) }
};

export default connect(mapStateToProps)(ExpenseList);
