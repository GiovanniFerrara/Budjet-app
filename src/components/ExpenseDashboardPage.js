import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseList />
    <ExpenseListFilter />
    <button onClick={() => {
      console.log(process.env.FIREBASE_API_KEY)
    }}>click</button>
  </div>
);

export default ExpenseDashboardPage;
