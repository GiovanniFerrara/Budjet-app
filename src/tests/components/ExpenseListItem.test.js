import React, { Component } from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses';

test('render ExpenseListItem shapshot with expenses', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot()
})