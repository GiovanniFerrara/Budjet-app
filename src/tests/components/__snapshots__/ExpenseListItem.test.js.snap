// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseListItem shapshot with expenses 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="edit/1"
    >
      Gum
    </Link>
  </h3>
  <p>
     
    1.95
     -  
    01-January-1970
     
  </p>
</div>
`;
